### PR TITLE
Updated lifecycle triggers as proposed.

### PR DESCRIPTION
**Pull request by DeathByTape**
#First attempt at resolving: https://github.com/yahoo/elide/issues/291

Critical for Elide 3.0 usability.

Still up for discussion:

Should we have an `@OnDelete` hook? I currently removed it, but we could similarly pass the current HttpRequest method into the `RequestScope` and then `@OnDelete` could be covered by `@PostCommit` or similar. Thoughts?

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

**Comment by DeathByTape** [comment]
NOTE: I am leaning toward the second approach for handling `@OnDelete` since `OnDelete` doesn't really mesh well with our mental model of how these lifestyle hooks are laid out in this new form. Namely, we don't care what the action is (i.e. `GET`, `DELETE`, etc.) but instead we care about when we're executing security checks and updating the datastore. I think allowing the triggered code to perform based on method is more appropriate.

**Comment by aklish** [comment]
There are things I like about the new annotations, but the lifecycle seems devoid of knowing what CRUD action was performed on the entity.  I'm not sure how well this new proposal will map to existing uses of the old annotations.

**Comment by DeathByTape** [comment]
@aklish: see my comment. If we push this information down into the `RequestScope`, we can pass that information down to the developer.

File: elide-annotations/src/main/java/com/yahoo/elide/annotation/OnCreate.java [hunk 1]
```diff
@@ -1,23 +1,23 @@
...
+ * The invoked function takes a RequestScope as parameter.
+ * @see com.yahoo.elide.security.RequestScope
  */
-@Target({METHOD, FIELD})
-@Retention(RUNTIME)
+@Target({ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
 public @interface OnCreate {
-
+    String value() default "";
```
Comment (aklish) [inline elide-annotations/src/main/java/com/yahoo/elide/annotation/OnCreate.java]: Not sure we need a value.

File: elide-annotations/src/main/java/com/yahoo/elide/annotation/PostCreate.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * This hook will be triggered <em>after</em> all security checks have been run and <em>after</em> the datastore
+ * has been committed.
+ *
+ * The invoked function takes a RequestScope as parameter.
+ * @see com.yahoo.elide.security.RequestScope
+ */
+@Target({ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface PostCreate {
+    String value() default "";
```
Comment (aklish) [inline elide-annotations/src/main/java/com/yahoo/elide/annotation/PostCreate.java]: Do we need this?

File: elide-annotations/src/main/java/com/yahoo/elide/annotation/PreCreate.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * This hook will be triggered <em>after</em> all security checks have been run, but <em>before</em> the datastore
+ * has been committed.
+ *
+ * The invoked function takes a RequestScope as parameter.
+ * @see com.yahoo.elide.security.RequestScope
+ */
+@Target({ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface PreCreate {
+    String value() default "";
```
Comment (aklish) [inline elide-annotations/src/main/java/com/yahoo/elide/annotation/PreCreate.java]: Do we need this?

File: elide-core/src/main/java/com/yahoo/elide/Elide.java [hunk 4]
```diff
@@ -261,12 +261,14 @@ public ElideResponse get(
             GetVisitor visitor = new GetVisitor(requestScope);
             Supplier<Pair<Integer, JsonNode>> responder = visitor.visit(parse(path));
             transaction.preCommit();
+            requestScope.runQueuedOnTriggers();
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/Elide.java]: Do we have the order of 'on' and 'pre' swapped here?

File: elide-core/src/main/java/com/yahoo/elide/Elide.java [hunk 5]
```diff
@@ -261,12 +261,14 @@ public ElideResponse get(
...
             requestScope.getPermissionExecutor().executeCommitChecks();
             transaction.flush(requestScope);
             ElideResponse response = buildResponse(responder.get());
+            requestScope.runQueuedPreTriggers();
             auditLogger.commit();
             transaction.commit(requestScope);
-            requestScope.runCommitTriggers();
+            requestScope.runQueuedPostTriggers();
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/Elide.java]: Should post run immediately before or after commit?

File: elide-core/src/main/java/com/yahoo/elide/core/RequestScope.java [hunk 6]
```diff
@@ -353,21 +374,87 @@ else if (permissionFilter.isPresent()) {
...
-        commitTriggers.add(() -> resource.runTriggers(OnCommit.class, fieldName));
+    /**
+     * Queue triggers for a particular resource and its field.
+     *
+     * @param resource Resource on which to execute trigger
+     * @param fieldName Field name for which to specify trigger
+     * @param crudAction CRUD Action
+     */
+    protected void queueTriggers(PersistentResource resource, String fieldName, CRUDAction crudAction) {
+        Consumer<Class> queue = (cls) -> queuedTriggers.get(cls).add(() -> resource.runTriggers(cls, fieldName));
```
Comment (clayreimann) [inline elide-core/src/main/java/com/yahoo/elide/core/RequestScope.java]: `querer` perhaps? `triggerQueuer`? `triggerFinder`?

**Comment by clayreimann** [comment]
👍

**Pull request merged**